src/test_fixture/event_gen.rs

```diff
@@ -0,0 +1,299 @@
+        debug_assert!(self.generated < self.max);
+
+        self.generated += 1;
+        self.generated == self.max
+    }
+}
+
+/// Generates random source and trigger events with guarantee that every next event
+/// occurs at the same time or after the previous event.
+///
+/// Number of events generated depends on the configured number of unique users per set
+/// and maximum number of events per user. See [`Config`] for more details
+///
+/// [`Config`]: Config
+pub struct EventGenerator<R: Rng> {
+    config: Config,
+    rng: R,
+    users: Vec<UserStats>,
+    // even bit vector takes too long to initialize for MAX_USER_ID. Need a sparse structure
+    // here
+    used: HashSet<UserId>,
+    current_ts: u64,
+}
+
+impl<R: Rng> EventGenerator<R> {
+    pub fn with_default_config(rng: R) -> Self {
+        Self::with_config(rng, Config::default())
+    }
+
+    pub fn with_config(rng: R, config: Config) -> Self {
+        Self {
+            config,
+            rng,
+            users: vec![],
+            used: HashSet::new(),
+            current_ts: 0,
+        }
+    }
+
+    fn gen_event(&mut self, user_id: UserId) -> TestRawDataRecord {
+        if self.rng.gen() {
+            self.current_ts += 1;
+        }
+
+        if self.rng.gen() {
+            self.gen_trigger(user_id)
+        } else {
+            self.gen_source(user_id)
+        }
+    }
+
+    fn gen_trigger(&mut self, user_id: UserId) -> TestRawDataRecord {
+        let trigger_value = self.rng.gen_range(1..self.config.max_trigger_value.get());
+
+        TestRawDataRecord {
+            user_id: user_id.into(),
+            timestamp: self.current_ts,
+            is_trigger_report: true,
+            breakdown_key: 0,
+            trigger_value,
+        }
+    }
+
+    fn gen_source(&mut self, user_id: UserId) -> TestRawDataRecord {
+        let breakdown_key = self.rng.gen_range(1..self.config.max_breakdown_key.get());
+
+        TestRawDataRecord {
+            user_id: user_id.into(),
+            timestamp: self.current_ts,
+            is_trigger_report: false,
+            breakdown_key,
+            trigger_value: 0,
+        }
+    }
+
+    fn sample_user(&mut self) -> Option<UserStats> {
+        if self.used.len() == self.config.max_user_id() + 1 {
+            return None;
+        }
+
+        let valid = |user_id| -> bool { !self.used.contains(&user_id) };
+
+        Some(loop {
+            let next = UserId::from(self.rng.gen_range(0..=self.config.max_user_id.get()));
+            if valid(next) {
+                self.used.insert(next);
+                break UserStats::new(
+                    next,
+                    self.rng
+                        .gen_range(1..=self.config.max_events_per_user.get()),
+                );
+            }
+        })
+    }
+}
+
+impl<R: Rng> Iterator for EventGenerator<R> {
+    type Item = TestRawDataRecord;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        const USERS_IN_FLIGHT: usize = 10;
```

Comment:
This seems a little small for larger event counts.  It's arbitrary, but I'd have said that you would expect hundreds of people for a million records.  (It's no extra work for IPA, but I'm not sure how much we care about fidelity.)